### ISSUE TITLE
About Confluence Macros

### ISSUE THREAD

**Issue by Skeeve** [justmiles/go-markdown2confluence#28]
I haven't had the time to check all the macros listed on https://confluence.atlassian.com/doc/macros-139387.html but I think all of those having **Macro body: none** will work with my enhancement.

I started looking into another enhancement which would support some or all of the macros having **Macro body: richt text** like so:

```markdown
    # This is an ugly test

    ```CONFLUENCE-MACRO
    name:panel
      borderColor:pink
      bgColor:green
      titleColor:red
      borderWidth:4
      titleBGColor:magenta
      borderStyle:dashed
      title:PanelTitle
    :rich-text-body
    ```

    Lorem ipsum dolor

    situs vilate inis et abanit

    ```CONFLUENCE-MACRO
    /rich-text-body
    ```

    # FIN
```

which would result in this Confluence Page:

```html
<h1>This is an ugly test</h1>
<ac:structured-macro ac:name="panel" ac:schema-version="1" ac:macro-id="3f76f3e8-51a3-4233-94c9-2c70c322b5ce">
  <ac:parameter ac:name="borderColor">pink</ac:parameter>
  <ac:parameter ac:name="bgColor">green</ac:parameter>
  <ac:parameter ac:name="titleColor">red</ac:parameter>
  <ac:parameter ac:name="borderWidth">4</ac:parameter>
  <ac:parameter ac:name="titleBGColor">magenta</ac:parameter>
  <ac:parameter ac:name="borderStyle">dashed</ac:parameter>
  <ac:parameter ac:name="title">PanelTitle</ac:parameter>
  <ac:rich-text-body>
    <p>Lorem ipsum dolor</p>
    <p>situs vilate inis et abanit</p>
  </ac:rich-text-body>
</ac:structured-macro>
<h1>FIN</h1>
```

It feels even more ugly than my previous version ;) 

Maybe we should discuss how to go forward from here?

**Comment by justmiles** [comment]
Yeah that definitely adds more complexity. The structure of that text already looks/reads like YAML. What if we just processed it as YAML. You could then have a generic multi-line field for the body like `rich-text-body` and all of that content is then rendered to xhtml. We'd have to actually process it as yaml, but then it could look something like this:

    ```CONFLUENCE-MACRO
    name: panel
    borderColor: pink
    bgColor: green
    titleColor: red
    borderWidth: 4
    titleBGColor: magenta
    borderStyle: dashed
    title: PanelTitle
    body: |
      ## Header inside panel
        
      More content inside panel
        
      - item 1
      - item 2
    ```

**Comment by justmiles** [comment]
Supporting Confluence's current markup format for macros would solve for this too. Instead of wrapping it in a code block we might need to add a first-class node for the macros like `{toc:...`, `{panel:...`, `{anchor:..` etc. With the end goal of it looking something like this:

    # My Page

    Some content outside of a macro

    {panel:title=PanelTitle|borderColor=pink|bgColor=green|titleColor=red|borderWidth=4|titleBGColor=magenta|borderStyle=dashed}
    ## Header inside panel
      
    More content inside panel
      
    - item 1
    - item 2
    {panel}

**Comment by justmiles** [comment]
At the end of the day I think most of the folks that find this useful are trying to sync content that is pure markdown to Confluence. If the community really needs the full suite of macros I'm sure we can make it work, but I think what you have from #27 covers the majority of the macro needs.

**Comment by Skeeve** [comment]
The issues I see with supporting the "Confluence Syntax" as a first-class element are:

1. the possibility of it clashing with future markdown
2. it's very inconvenient to read

 Because # 1 I choose to put it into a fenced code block. No chance at all it will ever clash with future markdown unless fenced code blocks get dropped.

I choose the yaml-like syntax because it's most easy to parse.

Of course we could parse it as real yaml, but this will open up even more complexity when it comes to parsing values as markdown.

Maybe in that case it would really be better to have new first-class elements.

Let's wait, whether or not the need will arise.
